### PR TITLE
(maint) Add the concept of a PDK Context

### PR DESCRIPTION
**Pull request by glennsarti**
This RFC is for adding the concept of a PDK context.

### CODE REVIEW

File: RFCs/0000-add-pdk-context.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+The combination of `root_path` and `context_path` allows arbitrary contexts to be built for testing purposes and will probably not be used outside of that scenario.
+
+Subclasses
+
+| Context | Description |
+|--------| ------------|
+| `PDK::Context::None` | Represents a PDK Context where it's not in anything, for example a random directory before running `pdk new module` |
+| `PDK::Context::Module` | Represents a context of being in a Puppet Module. |
+| `PDK::Context::ControlRepo` (Future) | Represents a context of being in a Puppet Directory Based Control Repository |
```
Comment (scotje) [inline RFCs/0000-add-pdk-context.md]: Would the behavior of validators and such then be expected to switch based on `if context.is_a?(PDK::Context::Module)` or some such?

Comment (glennsarti) [reply RFCs/0000-add-pdk-context.md]: Yes.

I did consider using a delegation/forwarder pattern, but I think that using subclassing means we can add additional methods e.g. The ControlRepo context would have a `module_paths` method which would list all the places a module could be located.  This wouldn't make sense on a `PDK::Context::Module` class (or `::None` for that matter)

Potentially you could use a `Context.type == :module` but, they just seems an additional layer where the object type would be fine?

**Comment by glennsarti** [comment]
> There may be some minor revisions to the implementation once we start to use it in anger.

Agreed.  I don't know what I don't know, yet!

**Comment by glennsarti** [comment]
Will merge next week if no other comments

**Pull request merged**